README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cccapi <img src="man/figures/logo.svg" align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/RaoulWolf/cccapi/workflows/R-CMD-check/badge.svg)](https://github.com/RaoulWolf/cccapi/actions)
[![Codecov test
coverage](https://codecov.io/gh/RaoulWolf/cccapi/branch/master/graph/badge.svg)](https://app.codecov.io/gh/RaoulWolf/cccapi?branch=master)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

> A [ZeroPM](https://zeropm.eu/) R package

The goal of the cccapi package is to provide a minimal and lightweight
interface to the [CAS Common
Chemistry](https://commonchemistry.cas.org/) API services. To get access
to the API services, you need to register at
<https://www.cas.org/services/commonchemistry-api>.

The dependencies of cccapi are kept at a bare minimum:
[curl](https://cran.r-project.org/web/packages/curl/index.html) for
handling the API requests and
[jsonlite](https://cran.r-project.org/web/packages/jsonlite/index.html)
to parse data from JSON format.

## Installation

You can install the development version of cccapi from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("ZeroPM-H2020/cccapi")
```

## Examples

The following examples show the functionality of the three API functions
of cccapi: `get_detail()`, `get_export()`, and `get_search()`.

This is an example which shows you how to get the record details for a
CAS Registry Number. In this case, for
[atrazine](https://en.wikipedia.org/wiki/Atrazine):

``` r
library(cccapi)
atz_det <- get_detail("1912-24-9")
str(atz_det)
#> List of 15
#>  $ uri                   : chr "substance/pt/1912249"
#>  $ rn                    : chr "1912-24-9"
#>  $ name                  : chr "Atrazine"
#>  $ image                 : chr "<svg width=\"181.11\" viewBox=\"0 0 6037 4324\" text-rendering=\"auto\" stroke-width=\"1\" stroke-opacity=\"1\""| __truncated__
#>  $ inchi                 : chr "InChI=1S/C8H14ClN5/c1-4-10-7-12-6(9)13-8(14-7)11-5(2)3/h5H,4H2,1-3H3,(H2,10,11,12,13,14)"
#>  $ inchiKey              : chr "InChIKey=MXWJVTOOROXGIU-UHFFFAOYSA-N"
#>  $ smile                 : chr "N(C(C)C)C=1N=C(NCC)N=C(Cl)N1"
#>  $ canonicalSmile        : chr "ClC=1N=C(N=C(N1)NC(C)C)NCC"
#>  $ molecularFormula      : chr "C<sub>8</sub>H<sub>14</sub>ClN<sub>5</sub>"
#>  $ molecularMass         : chr "215.68"
#>  $ experimentalProperties:'data.frame':  2 obs. of  3 variables:
#>   ..$ name        : chr [1:2] "Melting Point" "Density"
#>   ..$ property    : chr [1:2] "175-177 Â°C" "1.23 g/cm<sup>3</sup> @ Temp: 22 Â°C"
#>   ..$ sourceNumber: int [1:2] 1 1
#>  $ propertyCitations     :'data.frame':  1 obs. of  3 variables:
#>   ..$ docUri      : chr ""
#>   ..$ sourceNumber: int 1
#>   ..$ source      : chr "Hazardous Substances Data Bank data were obtained from the National Library of Medicine (US)"
#>  $ synonyms              : chr [1:71] "1,3,5-Triazine-2,4-diamine, 6-chloro-<em>N</em><sup>2</sup>-ethyl-<em>N</em><sup>4</sup>-(1-methylethyl)-" "<em>s</em>-Triazine, 2-chloro-4-(ethylamino)-6-(isopropylamino)-" "1,3,5-Triazine-2,4-diamine, 6-chloro-<em>N</em>-ethyl-<em>N</em>â\200²-(1-methylethyl)-" "6-Chloro-<em>N</em><sup>2</sup>-ethyl-<em>N</em><sup>4</sup>-(1-methylethyl)-1,3,5-triazine-2,4-diamine" ...
#>  $ replacedRns           : chr [1:6] "11121-31-6" "12040-45-8" "12797-72-7" "39400-72-1" ...
#>  $ hasMolfile            : logi TRUE
```

Here is an example which shows you how to get the MOL file for a CAS
Registry Number. In this case, again, for atrazine:

``` r
atz_exp <- get_export("1912-24-9")
str(atz_exp)
#>  chr "Atrazine\nC8H14ClN5\n1912-24-9 Copyright (C) 2021 ACS\n 14 14  0  0  0  0  0  0  0  0999 V2000\n    8.0021    0"| __truncated__
```

And finally, here is an example which shows you how to search for CAS
Registry Numbers with a partial name. See `?get_search` for details on
the available search options!

``` r
atz_src <- get_search("atrazin*")
str(atz_src)
#> List of 2
#>  $ count  : int 12
#>  $ results:'data.frame': 12 obs. of  3 variables:
#>   ..$ rn   : chr [1:12] "1007-28-9" "1912-24-9" "2163-68-0" "39324-65-7" ...
#>   ..$ name : chr [1:12] "Deisopropylatrazine" "Atrazine" "Hydroxyatrazine" "1,3,5-Triazine-2,4-diamine, 6-chloro-<em>N</em>-ethyl-<em>N</em>â\200²-(1-methylethyl)-, mixt. with <em>N</em>-"| __truncated__ ...
#>   ..$ image: chr [1:12] "<svg width=\"178.77\" viewBox=\"0 0 5959 4052\" text-rendering=\"auto\" stroke-width=\"1\" stroke-opacity=\"1\""| __truncated__ "<svg width=\"181.11\" viewBox=\"0 0 6037 4324\" text-rendering=\"auto\" stroke-width=\"1\" stroke-opacity=\"1\""| __truncated__ "<svg width=\"178.77\" viewBox=\"0 0 5959 4324\" text-rendering=\"auto\" stroke-width=\"1\" stroke-opacity=\"1\""| __truncated__ "" ...
```

## Acknowledgement

This R package was developed by the EnviData initiative at the
[Norwegian Geotechnical Institute (NGI)](https://www.ngi.no/eng) as part
of the project [ZeroPM: Zero pollution of Persistent, Mobile
substances](https://zeropm.eu/). This project has received funding from
the European Union’s Horizon 2020 research and innovation programme
under grant agreement No 101036756.

------------------------------------------------------------------------

If you find this package useful and can afford it, please consider
making a donation to a humanitarian non-profit organization, such as
[Sea-Watch](https://sea-watch.org/en/). Thank you.


NEWS.md
# cccapi 0.1.5 (2023-02-27)
* Non-200 status codes will now return the entire JSON string when 
  `json = TRUE`.  

# cccapi 0.1.4 (2022-10-04)

* Added functionality to `get_search()` to handle non-200 request statuses.

# cccapi 0.1.3 (2022-04-11)

* Vectorized the `.check_cas()` function.  

# cccapi 0.1.2 (2022-03-22)

* Safer queries in `get_search()` *via* `utils::URLencode()`.
* Updated "Get started" vignette. 
* Minor (non-functional) code cosmetics.

# cccapi 0.1.1 (2022-01-26)

* Fixed a bug when searching for SMILES with URL-problematic characters. 

# cccapi 0.1.0 (2022-01-19)

* Minor bugfixes.

# cccapi 0.0.4 (2022-01-07)

* Added logo to GitHub README.
* Removed ghost parameters from `get_detail()` and `get_search()`.
* Updated tests to reflect the removal of ghost parameters in `get_detail()` 
  and `get_search()`.

# cccapi 0.0.3 (2022-01-07)

* Added logical `json` parameter to `get_detail()` and `get_search()` to 
  retrieve the original JSON as returned from the API (*i.e.*, skipping the 
  `jsonlite::from_json()` parsing). The default value is `FALSE`. 
* Added the "Get Started with cccapi" vignette to demonstrate a basic workflow.

# cccapi 0.0.2 (2022-01-06)

* Basic support for the three main functions: `get_detail()`, `get_export()`
  and `get_search()`. 
* Implemented internal functionality to sanity-check CAS Registry Numbers 
  before running queries: `.check_cas()`.  
* Added continuous integration *via* GitHub Actions.
* Added tests via {tinytest} and added it to suggestions.
* Added coverage via {covr} and added it to suggestions.
* Added {magick} and {rsvg} to suggestions to work with SVG files. 
* Added life cycle badge

# cccapi 0.0.1 (2022-01-05)

* Added a `NEWS.md` file to track changes to the package.


tests/tinytest.R
#'
if (requireNamespace("tinytest", quietly = TRUE)) {
  tinytest::test_package("cccapi")
}


inst/tinytest/test_cccapi.R

# test get_detail()

expect_equal(
  length(unlist(get_detail(NA))),
  1L
  )

expect_equal(
  length(get_detail("50-78-2")),
  15L
)

expect_equal(
  typeof(get_detail("50-78-2", json = TRUE)),
  "character"
)

expect_equal(
  length(get_detail("50-78-2", json = "no")),
  15L
)

# test get_export()

expect_equal(
  nchar(get_export(NA)),
  28L
)

expect_equal(
  nchar(get_export("50-78-2")),
  1289L
)

# test get_search()

expect_equal(
  length(unlist(get_search(NA))),
  1L
)

expect_equal(
  length(unlist(get_search(NULL))),
  1L
)

expect_equal(
  length(get_search("Aspirin")),
  2L
)

expect_equal(
  length(get_search("Aspirin", json = "no")),
  2L
)

expect_equal(
  typeof(get_search("Aspirin", json = TRUE)),
  "character"
)

# test .check_cas()

expect_false(
  cccapi:::.check_cas(NA)
)

expect_false(
  cccapi:::.check_cas("cas_rn")
)

expect_false(
  cccapi:::.check_cas("50-78-1.1")
)

expect_false(
  cccapi:::.check_cas("50-78-")
)

expect_false(
  cccapi:::.check_cas("12345650-78-2")
)

expect_false(
  cccapi:::.check_cas("50-78-1")
)

expect_true(
  cccapi:::.check_cas("50-78-2")
)


R/check_cas.R
#' @source
#'   https://www.cas.org/support/documentation/chemical-substances/checkdig
.check_cas <- function(cas_rn, use_names = FALSE) {

  if (!is.logical(use_names) || is.na(use_names)) {
    use_names <- FALSE
  }

  res <- sapply(
    cas_rn,
    FUN = function(x) {

      if (is.na(x) || is.null(x)) {
        return(FALSE)
      }

      if (nchar(gsub(pattern = "[^-]", replacement = "", x)) != 2) {
        return(FALSE)
      }

      if (
        any(
          sapply(
            unlist(strsplit(x, split = "-")),
            FUN = function(x) { nchar(x) }
          ) == 0L
        )
	    ) {
        return(FALSE)
      }

      cas <- gsub(pattern = "-", replacement = "", x)

      if (grepl(pattern = "[^[:digit:]]", cas)) {
        return(FALSE)
      }

      if (nchar(cas) < 5 || nchar(cas) > 10) {
        return(FALSE)
      }

      check_dig <- as.integer(substring(cas, first = nchar(cas)))

      cas_str <- substr(cas, start = 1, stop = nchar(cas) - 1)

      count <- nchar(cas_str):1

      number <- as.integer(unlist(strsplit(cas_str, split = "")))

      res <- sum(count * number) / 10

      res_dig <- as.integer(round(10 * (res %% 1)))

      res_dig == check_dig

    },
    USE.NAMES = use_names
  )

  res

}


R/get_detail.R
#' @title GET Details for a CAS Registry Number
#' @description This function performs a query to retrieve details for a CAS
#'   Registry Number.
#' @param cas_rn (Character) CAS Registry Number with dashes, e.g., "50-78-2".
#' @param json (Logical) Should the result be returned as JSON? Defaults to
#'   \code{FALSE}.
#' @details The function performs a sanity check on the provided CAS Registry
#'   Number and then performs a query. If successful, a list with the available
#'   details will be returned.
#' @return Returns a list.
#' @author Raoul Wolf (\url{https://github.com/RaoulWolf/})
#' @examples \dontrun{
#' cas_rn <- "50-78-2"
#' get_detail(cas_rn)
#' }
#' @importFrom curl curl_fetch_memory handle_setopt new_handle
#' @importFrom jsonlite fromJSON
#' @export
get_detail <- function(cas_rn, json = FALSE) {

  # sanity-check cas_rn
  if (isFALSE(.check_cas(cas_rn))) {
    return(list("message" = "Invalid CAS Registry Number."))
  }

  # sanity-check json
  if (!is.logical(json) || is.na(json)) {
    json <- FALSE
  }

  # compose url
  url <- paste0("commonchemistry.cas.org/api/detail?cas_rn=", cas_rn)

  # create new cURL handle
  handle <- curl::new_handle()

  # specify GET request
  curl::handle_setopt(handle, customrequest = "GET")

  # retrieve results
  result <- curl::curl_fetch_memory(url, handle)

  # decode content
  content <- rawToChar(result$content)

  # check status
  if(result$status_code != 200L) {
    if(json) {
      return(content)
    } else {
      warning("Detail not found", call. = FALSE)
      return(list())
    }
  }

  # transform content
  if(!json) {
    content <- jsonlite::fromJSON(content)
  }

  # return content
  content

}


R/get_search.R
#' @title GET Search Results for a CAS Registry Number and Other Identifiers
#' @description This function performs a query to search for substances based
#'   on a CAS Registry Number or other identifiers.
#' @param q (Character) The (partial) search term. The following search terms
#'   are supported: CAS Registry Number (with or without dashes), SMILES
#'   (canonical or isomeric), InChI (with or without the "InChI=" prefix),
#'   InChIKey, and name. Searching by name allows use of a trailing wildcard
#'   (e.g., car*). All searches are case-insensitive.
#' @param json (Logical) Should the result be returned as JSON? Defaults to
#'   \code{FALSE}.
#' @details The function performs a sanity check on the provided search term
#'   and then performs a query. If successful, a list with the available
#'   information will be returned.
#' @note When searching for an InChIKey, the prefix \code{"InChIKey="} needs to
#'   be added.
#' @return Returns a list.
#' @author Raoul Wolf (\url{https://github.com/RaoulWolf/})
#' @examples \dontrun{
#' q <- "Aspiri*"
#' get_search(q)
#' }
#' @importFrom curl curl_fetch_memory handle_setopt new_handle
#' @importFrom jsonlite fromJSON
#' @importFrom utils URLencode
#' @export
get_search <- function(q, json = FALSE) {

  # sanity-check q
  if (missing(q) || is.na(q) || is.null(q)) {
    return(list("message" = "Invalid search term."))
  }

  # sanity-check json
  if (!is.logical(json) || is.na(json)) {
    json <- FALSE
  }

  # compose url
  url <- paste(
    "https://commonchemistry.cas.org/api/search",
    paste("q", utils::URLencode(q, reserved = TRUE), sep = "="),
    sep = "?"
  )

  # create new cURL handle
  handle <- curl::new_handle()

  # specify GET request
  curl::handle_setopt(handle, customrequest = "GET")

  # retrieve results
  result <- curl::curl_fetch_memory(url, handle)

  # decode content
  content <- rawToChar(result$content)

  # check status
  if(result$status_code != 200L) {
    if(json) {
      return(content)
    } else {
      return(list())
    }
  }

  # transform content
  if(!json) {
    content <- jsonlite::fromJSON(content)
  }

  # return content
  content

}


R/get_export.R
#' @title GET the MOL Export for a CAS Registry Number
#' @description This function performs a query to retrieve a MOL file for a CAS
#'   Registry Number.
#' @param cas_rn (Character) CAS Registry Number in "dash" format, e.g.,
#'   "50-78-2".
#' @param verbose (Logical) Should the result be returned as is? Defaults to
#'   \code{FALSE}.
#' @details The function performs a sanity check on the provided CAS Registry
#'   Number and then performs a query. If successful, a character string with
#'   the available MOL file will be returned.
#' @return Returns a character string.
#' @author Raoul Wolf (\url{https://github.com/RaoulWolf/})
#' @examples \dontrun{
#' cas_rn <- "50-78-2"
#' get_export(cas_rn)
#' }
#' @importFrom curl curl_fetch_memory handle_setopt new_handle
#' @export
get_export <- function(cas_rn, verbose = FALSE) {

  # sanity-check cas_rn
  if (isFALSE(.check_cas(cas_rn))) {
    return("Invalid CAS Registry Number.")
  }

  # remove cas_rn dashes
  cas <- gsub(pattern = "-", replacement = "", cas_rn)

  # construct uri
  uri <- paste0("substance/pt/", cas)

  # define attachment
  attachment <- paste0("&returnAsAttachment=", tolower(FALSE))

  # compose url
  url <- paste0("commonchemistry.cas.org/api/export?uri=", uri, attachment)

  # create new cURL handle
  handle <- curl::new_handle()

  # specify GET request
  curl::handle_setopt(handle, customrequest = "GET")

  # retrieve results
  result <- curl::curl_fetch_memory(url, handle)

  # decode content
  content <- rawToChar(result$content)

  # check status
  if(result$status_code != 200L) {
    if(verbose) {
      return(content)
    } else {
      return(NA_character_)
    }
  }

  # return content
  content

}
